domain/storage/state/storagepool.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	domainstorage "github.com/juju/juju/domain/storage"
+	storageerrors "github.com/juju/juju/domain/storage/errors"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// StoragePoolState represents database interactions dealing with storage pools.
+type StoragePoolState struct {
+	*domain.StateBase
+}
+
+type poolAttributes map[string]string
+
+// CreateStoragePool creates a storage pool, returning an error satisfying [storageerrors.PoolAlreadyExists]
+// if a pool with the same name already exists.
+func (st StoragePoolState) CreateStoragePool(ctx context.Context, pool domainstorage.StoragePoolDetails) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return CreateStoragePools(ctx, db, []domainstorage.StoragePoolDetails{pool})
+}
+
+// CreateStoragePools creates the specified storage pools.
+// It is exported for us in the storage/bootstrap package.
+func CreateStoragePools(ctx context.Context, db coredatabase.TxnRunner, pools []domainstorage.StoragePoolDetails) error {
+	selectUUIDStmt, err := sqlair.Prepare("SELECT &StoragePool.uuid FROM storage_pool WHERE name = $StoragePool.name", StoragePool{})
+	if err != nil {
+		return errors.Trace(domain.CoerceError(err))
+	}
+
+	storagePoolUpserter, err := storagePoolUpserter()
+	if err != nil {
+		return errors.Trace(domain.CoerceError(err))
+	}
+	poolAttributesUpdater, err := poolAttributesUpdater()
+	if err != nil {
+		return errors.Trace(domain.CoerceError(err))
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		for _, pool := range pools {
+			dbPool := StoragePool{Name: pool.Name}
+			err := tx.Query(ctx, selectUUIDStmt, dbPool).Get(&dbPool)
+			if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Trace(domain.CoerceError(err))
+			}
+			if err == nil {
+				return fmt.Errorf("storage pool %q %w", pool.Name, storageerrors.PoolAlreadyExists)
+			}
+			poolUUID := uuid.MustNewUUID().String()
```

Comment:
Make this outside of the for-loop. We're creating garbage for no reason.

Reply:
I don't follow sorry - the for loop is for a set of new pools to create. Each needs its own uuid.

Reply:
yes